### PR TITLE
Introduce assertions

### PR DESCRIPTION
**Pull request by lgalabru**
Address #1124 

I'm not sure if the type checker should be yelling when checking the following expression:
```
(begin (asserts! (eq? 1 0) 'false) 1)
```
since the execution would be throwin anyway.
Thoughts?

### CODE REVIEW

**Comment by kantai** [comment]
> I'm not sure if the type checker should be yelling when checking the following expression:
>
> (begin (asserts! (eq? 1 0) 'false) 1)
> since the execution would be throwin anyway.
> Thoughts?

I don't see why the type checker would complain about that snippet -- it seems like the functions are all being supplied with the correct types and argument counts.

File: src/vm/docs/mod.rs [hunk 1]
```diff
@@ -559,6 +559,16 @@ principal and executes `expr` with that context. It returns the resulting value
     example: "(as-contract (print tx-sender)) ;; Returns 'CTcontract.name"
 };
 
+const ASSERTS_API: SpecialAPI = SpecialAPI {
+    input_type: "(optional A) | (response A B), C",
```
Comment (kantai) [inline src/vm/docs/mod.rs]: input_type should be: `bool, C`

Comment (lgalabru) [reply src/vm/docs/mod.rs]: Addressed with https://github.com/blockstack/blockstack-core/pull/1127/commits/a921acfa2f20dcbab7e2b85adce3d1620bf0368e

File: src/vm/functions/mod.rs [hunk 2]
```diff
@@ -221,6 +223,25 @@ fn special_if(args: &[SymbolicExpression], env: &mut Environment, context: &Loca
...
+    // handle the conditional clause.
+    let conditional = eval(&args[0], env, context)?;
+
+    match conditional {
+        Value::Bool(result) => {
+            if result {
+                Ok(conditional)
+            } else {
+                let thrown = eval(&args[1], env, context)?;
+                Err(ShortReturnType::ExpectedValue(thrown.clone()).into())
```
Comment (jcnelson) [inline src/vm/functions/mod.rs]: Should this be an `asserts!`-specific error, like `ShortReturnType::AssertionFailed`?  I thought `ShortReturnType::ExpectedValue` was specific to `expects`?

Comment (lgalabru) [reply src/vm/functions/mod.rs]: Addressed with https://github.com/blockstack/blockstack-core/pull/1127/commits/514052911f4044cb0857677b21765ff44d1685e8

**Pull request merged**